jetty-core/jetty-core-server/src/main/java/org/eclipse/jetty/core/server/Handler.java

```diff
@@ -243,6 +267,11 @@ public Handler getHandler()
 
         public void setHandler(Handler handler)
         {
+            Server server = getServer();
+            if (server != null && server.isStarted() && handler != null &&
+                server.getInvocationType() != Invocable.combine(server.getInvocationType(), handler.getInvocationType()))
+                throw new IllegalArgumentException("Cannot change invocation type of started server");
```

Comment:
@sbordet we don't really want to recompute the runhandle for every channel, as if they are not reused (or connection is not long lived) then that is a lot of extra traverses of the full handler tree.
The ISE will only be thrown if the computed type is changed to blocking when it is cached as non-blocking.  So it should be rare.   But we could equally just recompute the type instead of throwing IAE

Reply:
Many folks have dynamic handler trees now, we should still allow for this kind of thing, even if it puts more burden on the developers using it.

Reply:
@joakime this doesn't prevent dynamic trees.  It only currently prevents dynamic trees that change the dispatch type... but as any app that has a servlet is already going to be BLOCKING, this will not affect them.     It would only affect users that created dynamic handler trees that are async and then added a blocking handler.  We have zero users like that now, because we don't have async handlers.

Reply:
@sbordet @joakime I've had a look at making the type dyanmic as well, but I don't think it can be done without making a global lock on all handling. Anytime you ask the handler tree for a type, then an async modification can sneak in and change the tree before you actually call the first handler.  Thus we are always in a race.

Choices are:  
 1. always assume handlers are blocking (as we currently do)
 2. Allow async handlers, but once we have started with a fully async tree, then we can't change it to blocking dynamically.  We can always add a async handler to a blocking tree, but not vice versa